src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,197 @@
+);
+
+
+typedef struct {
+    float max;
+    float min;
+    float kp;
+    float kd;
+    float ki;
+    float lastErr;
+    float integral;
+} simplePid_s;
+
+void simplePidInit(simplePid_s* simplePid, float min, float max, float kp, float kd, float ki)
+{
+    simplePid->max = max;
+    simplePid->min = min;
+    simplePid->kp = kp;
+    simplePid->kd = kd;
+    simplePid->ki = ki;
+    simplePid->lastErr = 0;
+    simplePid->integral = 0;
+}
+
+float simplePidCalculate(simplePid_s* simplePid, float dt, float targetValue, float measuredValue)
+{
+    float error = targetValue - measuredValue;
+
+    float pOut = simplePid->kp * error;
+
+    float iOut = simplePid->ki * simplePid->integral;
+
+    simplePid->integral += error * dt;
+
+    float derivative = (error - simplePid->lastErr) / dt;
+    float dOut = simplePid->kd * derivative;
+
+    float output = pOut + iOut + dOut;
+    output = constrainf(output, simplePid->min, simplePid->max);
+
+    simplePid->lastErr = error;
+    return output;
+}
+
+typedef struct {
+    simplePid_s altPid;
+    simplePid_s velPid;
+    float throttle;
+    float targetAltitude;
+    float measuredAltitude;
+    float measuredAccel;
+    float velocityEstimationAccel;  // based on acceleration
+    float startVelocityEstimationAccel;
+    bool prevAltHoldModeEnabled;
+} altHoldState_s;
+
+void altHoldReset(altHoldState_s* altHoldState)
+{
+    simplePidInit(&altHoldState->altPid, -50.0f, 50.0f,
+                  0.01f * altholdConfig()->altPidP,
+                  0.0f,
+                  0.01f * altholdConfig()->altPidI);
+
+    simplePidInit(&altHoldState->velPid, 0.0f, 1.0f,
+                  0.01f * altholdConfig()->velPidP,
+                  0.01f * altholdConfig()->velPidD,
+                  0.0f);
+
+    altHoldState->throttle = 0.0f;
+    altHoldState->startVelocityEstimationAccel = altHoldState->velocityEstimationAccel;
+    altHoldState->targetAltitude = (float)(0.01f * getEstimatedAltitudeCm());
+}
+
+void altHoldInit(altHoldState_s* altHoldState)
+{
+    altHoldState->prevAltHoldModeEnabled = false;
+    altHoldState->velocityEstimationAccel = 0.0f;
+    altHoldReset(altHoldState);
+}
+
+
+void altHoldUpdate(altHoldState_s* altHoldState)
+{
+    bool altHoldModeEnabled = FLIGHT_MODE(ALTHOLD_MODE);
+
+    if (altHoldModeEnabled && !altHoldState->prevAltHoldModeEnabled)
+    {
+        altHoldReset(altHoldState);
+    }
+    altHoldState->prevAltHoldModeEnabled = altHoldModeEnabled;
+
+
+    float measuredAltitude = (float)(0.01f * getEstimatedAltitudeCm());
+
+    t_fp_vector accelerationVector = {{
+        acc.accADC[X],
+        acc.accADC[Y],
+        acc.accADC[Z]
+    }};
+
+    imuTransformVectorBodyToEarth(&accelerationVector);
+
+    float measuredAccel = 9.8f * (accelerationVector.V.Z - acc.dev.acc_1G) / acc.dev.acc_1G;
+
+    DEBUG_SET(DEBUG_ALTHOLD, 0, (int16_t)(measuredAccel * 100.0f));
+
+    altHoldState->velocityEstimationAccel += measuredAccel * 0.01f;
```

Comment:
Actual time delta since last call should be used here (and in other places)

Reply:
I'm not sure that all this code is robust to variable time-frames (in fact, I'm almost sure that it doesn't). For simplicity we consider that everything works on constant rate, if it's not true everything becomes more complicated.

For a small mismatch there will not be much difference on practice, but on a large one - there definitely would.

Not sure that it's a good idea to spend much time on making and testing a perfect model here.

Reply:
Most places will be fine (velocity integration, PIDs). But there will be some sampling artefacts (IMU sample rate / althold invocation jitter) , hopefully not too important.
At least use #define for update rate, so that realtion is obvious

Reply:
Added define